cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/AbstractRackAwareGoal.java

```diff
@@ -0,0 +1,169 @@
+          return BROKER_REJECT;
+        }
+
+        if (action.balancingAction() == ActionType.INTER_BROKER_REPLICA_SWAP
+            && doesReplicaMoveViolateActionAcceptance(clusterModel,
+                                                      c -> c.broker(action.destinationBrokerId()).replica(action.destinationTopicPartition()),
+                                                      c -> c.broker(action.sourceBrokerId()))) {
+          return REPLICA_REJECT;
+        }
+        return ACCEPT;
+      default:
+        throw new IllegalArgumentException("Unsupported balancing action " + action.balancingAction() + " is provided.");
+    }
+  }
+
+  protected abstract boolean doesReplicaMoveViolateActionAcceptance(ClusterModel clusterModel,
+                                                                    Function<ClusterModel, Replica> sourceReplicaFunction,
+                                                                    Function<ClusterModel, Broker> destinationBrokerFunction);
+
+  /**
+   * Rebalance the given broker without violating the constraints of this custom rack aware goal and optimized goals.
+   *
+   * @param broker Broker to be balanced.
+   * @param clusterModel The state of the cluster.
+   * @param optimizedGoals Optimized goals.
+   * @param optimizationOptions Options to take into account during optimization.
+   * @param failIfCannotMove {@code true} to throw an {@link OptimizationFailureException} in case a required balancing
+   * action for a replica fails for all rack-aware eligible brokers, {@code false} to just log the failure and return.
+   * this parameter enables selected goals fail early in case the unsatisfiability of a goal can be determined early.
+   */
+  protected void rebalanceForBroker(Broker broker,
+                                    ClusterModel clusterModel,
+                                    Set<Goal> optimizedGoals,
+                                    OptimizationOptions optimizationOptions,
+                                    boolean failIfCannotMove)
+      throws OptimizationFailureException {
+    for (Replica replica : broker.trackedSortedReplicas(replicaSortName(this, false, false)).sortedReplicas(true)) {
+      if (broker.isAlive() && !broker.currentOfflineReplicas().contains(replica) && shouldKeepInTheCurrentRack(replica, clusterModel)) {
+        continue;
+      }
+      // The relevant rack awareness condition is violated. Move replica to an eligible broker
+      SortedSet<Broker> eligibleBrokers = rackAwareEligibleBrokers(replica, clusterModel);
+      if (maybeApplyBalancingAction(clusterModel, replica, eligibleBrokers,
+                                    ActionType.INTER_BROKER_REPLICA_MOVEMENT, optimizedGoals, optimizationOptions) == null) {
+        if (failIfCannotMove) {
+          throw new OptimizationFailureException(String.format("Failed to move replica %s to any broker in %s", replica, eligibleBrokers));
```

Comment:
Good question -- yes the user is made aware of the specific failure with details on how they may resolve them.
\--
An `OptimizationFailureException` causes the optimization process for requests, such as `rebalance`, `remove_broker`, or `add_broker,` to fail. 

Such failures and their relevant root cause (e.g. Insufficient disk capacity / number of racks, failure to move the offline replica to an alive broker) are logged and returned in response along with a generic potential mitigation strategy so that the user would know what likely action may help in resolving them (see `GoalUtils#mitigationForOptimizationFailures(OptimizationOptions)`).